# Contents List

## 01. Importing Libraries
## 02. Paths
## 03. Importing Data
## 04. User-defined Functions¶
## 05. The loc() Functions¶
## 06. Applying the loc() Function to the entire 'ords_prods_merge' dataframe¶
## 07. If-Statements with For-Loops
## Task 4.6 B

# 01. Importing libraries

In [2]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 2. Paths

In [4]:
#Creating a path varaible - Shortcut for Importing Data Files
path = r'/Users/aya/Desktop/DataAnalytics/Data-Immersion/Achievement-4/Instacart-Basket-Analysis'

In [5]:
path

'/Users/aya/Desktop/DataAnalytics/Data-Immersion/Achievement-4/Instacart-Basket-Analysis'

# 03. Importing Data

In [7]:
#Import ords_prods_merge Dataset
ords_prods_merge = pd.read_pickle(os.path.join(path, '02-Data', 'Prepared-Data', 'ords_prods_merge.pkl'))

In [8]:
#Checking the dataframe
ords_prods_merge

product_id                product_name  aisle_id  department_id  \
0                  1  Chocolate Sandwich Cookies        61             19   
1                  1  Chocolate Sandwich Cookies        61             19   
2                  1  Chocolate Sandwich Cookies        61             19   
3                  1  Chocolate Sandwich Cookies        61             19   
4                  1  Chocolate Sandwich Cookies        61             19   
...              ...                         ...       ...            ...   
32404854       49688      Fresh Foaming Cleanser        73             11   
32404855       49688      Fresh Foaming Cleanser        73             11   
32404856       49688      Fresh Foaming Cleanser        73             11   
32404857       49688      Fresh Foaming Cleanser        73             11   
32404858       49688      Fresh Foaming Cleanser        73             11   

          prices  order_id  user_id  order_number  orders_day_of_week  \
0            5.8   3139998      138            28                   6   
1            5.8   1977647      138            30                   6   
2            5.8    389851      709             2                   0   
3            5.8    652770      764             1                   3   
4            5.8   1813452      764             3                   4   
...          ...       ...      ...           ...                 ...   
32404854    13.5   1788356   200215             2                   0   
32404855    13.5   3401313   200377             1                   4   
32404856    13.5    809510   200873             5                   3   
32404857    13.5   2359893   200873             9                   3   
32404858    13.5   2385091   205926            11                   1   

          order_hour_of_day  days_since_prior_order  add_to_cart_order  \
0                        11                     3.0                  5   
1                        17                    20.0                  1   
2                        21                     6.0                 20   
3                        13                     NaN                 10   
4                        17                     9.0                 11   
...                     ...                     ...                ...   
32404854                  9                     5.0                 27   
32404855                 11                     NaN                  5   
32404856                  8                    15.0                 12   
32404857                 15                     5.0                 11   
32404858                 15                     6.0                 35   

          reordered _merge  
0                 0   both  
1                 1   both  
2                 0   both  
3                 0   both  
4                 1   both  
...             ...    ...  
32404854          0   both  
32404855          0   both  
32404856          0   both  
32404857          1   both  
32404858          0   both  

[32404859 rows x 14 columns]

# 04. User-defined Functions

In [10]:
#Create a subset of the data
df = ords_prods_merge[:1000000]

In [11]:
df.shape

(1000000, 14)

In [12]:
#Creating a function named "price_label"
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return np.nan

In [13]:
price_label

<function __main__.price_label(row)>

In [14]:
#Applying the function to the dataframe
df['price_range'] = df.apply(price_label, axis=1)

/var/folders/62/xltcnc5n7d97j6t5ql_87vym0000gn/T/ipykernel_13915/981087103.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [15]:
#Checking the function application on the dataframe
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    652638
Low-range product    338018
High range             9344
Name: count, dtype: int64

In [16]:
df['prices'].max()

24.5

# 05. The loc() Functions

In [18]:
#Creating conditions for for the loc() funtion
#Condition for High-range product
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

/var/folders/62/xltcnc5n7d97j6t5ql_87vym0000gn/T/ipykernel_13915/2748096239.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [19]:
#Creating conditions for for the loc() funtion
#Condition for Mid-range product
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [20]:
#Creating conditions for for the loc() funtion
#Condition for Low-range product
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [21]:
#Checking the function application on the dataframe
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     652638
Low-range product     338018
High-range product      9344
Name: count, dtype: int64

# 06. Applying the loc() Function to the entire 'ords_prods_merge' dataframe

In [23]:
#Creating conditions for for the loc() funtion
#Condition for High-range product
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [24]:
#Creating conditions for for the loc() funtion
#Condition for Mid-range product
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [25]:
#Creating conditions for for the loc() funtion
#Condition for Low-range product
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [26]:
#Checking the function application on the dataframe
ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21865987
Low-range product     10126321
High-range product      412551
Name: count, dtype: int64

In [27]:
#Checking the maximum prices in the 'ords_prods_merge' dataframe
ords_prods_merge['prices'].max()

25.0

In [28]:
#Check for missing values in the 'prices' column in the 'ords_prods_merge' dataframe
ords_prods_merge['prices'].isnull().sum()

0

In [29]:
ords_prods_merge.shape

(32404859, 15)

# 07. If-Statements with For-Loops

In [31]:
#printing the frequency of the “orders_day_of_week” - to find the busiest and slowest days of sales
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [32]:
#Using the for-loop to step through values in the "orders_day_of_week" column to categorize the values
result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [33]:
#Checking the 'result' list created
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

In [34]:
#Combining the 'result' with the 'ords_prods_merge' dataframe - creating a new column named 'busiest_day' in the 'ords_prods_merge' dataframe
ords_prods_merge['busiest_day'] = result

In [35]:
#Checking the frequency of the 'orders_day_of_week' in the context of the 'busiest_day' -the new column
ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# 08. Task 4.7

In [37]:
#Create a new column named 'Busiest_days' that contains three different values: Busiest days, Slowest days, & Regular days
result_2 = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_2.append("Busiest days")
  elif value == 4 or value == 3:
    result_2.append("Slowest days")
  else:
    result_2.append("Regular days")

In [38]:
result_2

['Regular days',
 'Regular days',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Regular days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Slowest days',
 'Slowest days',
 'Regular days',
 'Slowest days',
 'Regular days',
 'Regular days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Regular days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regular days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regular days',
 'Busiest days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Busiest days

In [39]:
#Combining the 'results_2' with the 'ords_prods_merge' dataframe, by
#Creating a new column whihtin the 'ords_prods_merge' dataframe named "Busiest_days"
ords_prods_merge['Busiest_days'] = result_2

In [40]:
#Checking the frequency of the 'orders_day_of_week' in the context of the 'Busiest days' -the new column
ords_prods_merge['Busiest_days'].value_counts(dropna = False)

Busiest_days
Regular days    12916111
Busiest days    11864412
Slowest days     7624336
Name: count, dtype: int64

In [41]:
ords_prods_merge.shape

(32404859, 17)

# Observation 2 & 3.
##### A new colimn was created named 'Busiest_days' in the ords_prods_merge dataframe
##### The column contains three categories:
##### 'Regular days' are the days when orders_day_of_week is: 2 (Monday), 5 (Thursday), or 6 (Friday).
##### 'Busiest days' are the days when orders_day_of_week is: 0 (Saturday), or 1 (Sunday).
##### 'Slowest days' are the days when orders_day_of_week is: 3 (Tuesday), or 4 (Wednesday).

##### The ords_prods_merge dataframe containes 32404859 rows (unchanged), 
##### However, the number of columns in the ords_prods_merge dataframe changed to 17. 
##### The new columns added are Busiest_days, busiest_day, and price_range_loc

In [43]:
#Finding the busiest hours of the day
busiest_hours_day = ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

In [44]:
busiest_hours_day

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [45]:
#Using the for-loop to step through values in the "order_hour_of_day" column to categorize the values
result_3 = []

for value in ords_prods_merge["order_hour_of_day"]:
    if value in (10, 11, 14, 15, 13, 12, 16, 9):
        result_3.append("Most orders")
    elif value in (23, 6, 0, 1, 5, 2, 4, 3):
        result_3.append("Fewest orders")
    else:
        result_3.append("Average orders")

In [46]:
result_3

['Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most ord

In [47]:
#Combining the 'results_3' with the 'ords_prods_merge' dataframe, by
#Creating a new column whihtin the 'ords_prods_merge' dataframe named "busiest_period_of_day"
ords_prods_merge['busiest_period_of_day'] = result_3

In [48]:
#Checking the frequency of the 'orders_day_of_week' in the context of the 'busiest_period_of_day days' -the new column
ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

In [49]:
#Checking the 'ords_prods_merge' dataframe
ords_prods_merge.shape

(32404859, 18)

# Observation 4. & 5.
#### The busiest hours of the day is 10. 

#### Creating three categories for the busiest hours of the day:
#### The values in the 'orders_day_of_week' column of the 'ords_prods_merge' dataframe are categorized as follows:
#### Most orders, Average orders, and Fewest orders.

#### The following cutoff for the orders is what I have decided makes sense
#### Most orders contain orders placed at hours 10, 11, 14, 15, 13, 12, 16, 9
#### The average orders contain orders placed at hours 17, 8, 18, 19, 20, 7, 21, 22
#### Fewest orders contain orders placed at hours 23, 6, 0, 1, 5, 2, 4, 3

#### Combined the result_3 list of data with the ords_prods_merge dataframe.
#### Created a new column named 'busiest_period_of_day' in the ords_prods_merge dataframe - the column values are the three new categories.

#### Checking the frequency of the 'orders_day_of_week' in the context of the 'busiest_period_of_day' - the new column.
#### The ords_prods_merge dataframe contains 32404859 rows, unchanged.
#### However, the number of columns in the ords_prods_merge dataframe has changed to 18.

#### A total of 4 new columns have been added: Busiest_days, busiest_day, price_range_loc, and busiest_period_of_day

In [51]:
#Export the ords_prods_merge dataframe to pkl formate
ords_prods_merge.to_pickle(os.path.join(path, '02-Data','Prepared-Data', 'ords_prods_merge_Deriving_Variables.pkl'))

### Observation 7.
##### Export the merged data, in pickle format as “ords_prods_merge_Deriving_Variables.pkl” 